core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -0,0 +1,783 @@
+}
+
+void hisi_qm_dev_init(struct hisi_qm *qm)
+{
+	if (qm->fun_type == HISI_QM_HW_VF)
+		return;
+
+	/* qm user domain */
+	io_write32(qm->io_base + QM_ARUSER_M_CFG_1, QM_AXUSER_CFG);
+	io_write32(qm->io_base + QM_ARUSER_M_CFG_ENABLE, AXUSER_M_CFG_ENABLE);
+	io_write32(qm->io_base + QM_AWUSER_M_CFG_1, QM_AXUSER_CFG);
+	io_write32(qm->io_base + QM_AWUSER_M_CFG_ENABLE, AXUSER_M_CFG_ENABLE);
+	/* qm cache */
+	io_write32(qm->io_base + QM_AXI_M_CFG, AXI_M_CFG);
+
+	if (qm->version == HISI_QM_HW_V2) {
+		/* disable FLR triggered by BME(bus master enable) */
+		io_write32(qm->io_base + QM_PEH_AXUSER_CFG, PEH_AXUSER_CFG);
+		/* set sec sqc and cqc cache wb threshold 4 */
+		io_write32(qm->io_base + QM_CACHE_CTL, QM_CACHE_CFG);
+	}
+	/* disable qm ras */
+	io_write32(qm->io_base + HISI_QM_ABNML_INT_MASK,
+		   HISI_QM_ABNML_INT_MASK_CFG);
+}
+
+static enum hisi_drv_status qm_sqc_cfg(struct hisi_qp *qp)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	struct hisi_qm *qm = qp->qm;
+	struct qm_sqc *sqc = NULL;
+	paddr_t sqc_dma = 0;
+
+	sqc = memalign(HISI_QM_ALIGN32, sizeof(struct qm_sqc));
+	if (!sqc)
+		return HISI_QM_DRVCRYPT_ENOMEM;
+
+	sqc_dma = virt_to_phys(sqc);
+
+	memzero_explicit(sqc, sizeof(struct qm_sqc));
+	reg_pair_from_64(qp->sqe_dma, &sqc->base_h, &sqc->base_l);
+	sqc->dw3 = (HISI_QM_Q_DEPTH - 1) |
+		    SHIFT_U32(qm->sqe_log2_size, QM_SQ_SQE_SIZE_SHIFT);
+	sqc->rand_data = QM_DB_RAND_DATA;
+	sqc->cq_num = qp->qp_id;
+	sqc->w13 = BIT32(QM_SQ_ORDER_SHIFT) |
+		   SHIFT_U32(qp->sq_type, QM_SQ_TYPE_SHIFT);
+
+	ret = qm_mb(qm, QM_MB_CMD_SQC, sqc_dma, qp->qp_id, QM_MB_OP_WR);
+	free(sqc);
+
+	return ret;
+}
+
+static enum hisi_drv_status qm_cqc_cfg(struct hisi_qp *qp)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	struct hisi_qm *qm = qp->qm;
+	struct qm_cqc *cqc = NULL;
+	paddr_t cqc_dma = 0;
+
+	cqc = memalign(HISI_QM_ALIGN32, sizeof(struct qm_cqc));
+	if (!cqc)
+		return HISI_QM_DRVCRYPT_ENOMEM;
+
+	cqc_dma = virt_to_phys(cqc);
+
+	memzero_explicit(cqc, sizeof(struct qm_cqc));
+	reg_pair_from_64(qp->cqe_dma, &cqc->base_h, &cqc->base_l);
+	cqc->dw3 = (HISI_QM_Q_DEPTH - 1) |
+		    SHIFT_U32(QM_CQE_SIZE, QM_CQ_CQE_SIZE_SHIFT);
+	cqc->rand_data = QM_DB_RAND_DATA;
+	cqc->dw6 = PHASE_DEFAULT_VAL;
+
+	ret = qm_mb(qm, QM_MB_CMD_CQC, cqc_dma, qp->qp_id, QM_MB_OP_WR);
+	free(cqc);
+
+	return ret;
+}
+
+struct hisi_qp *hisi_qm_create_qp(struct hisi_qm *qm, uint8_t sq_type)
+{
+	struct hisi_qp *qp = NULL;
+
+	mutex_lock(&qm->qp_lock);
+	if (qm->qp_in_used == qm->qp_num) {
+		EMSG("All %"PRIu32" queues of QM are busy", qm->qp_num);
+		goto err_proc;
+	}
+
+	if (qm->qp_idx == qm->qp_num - 1)
+		qm->qp_idx = 0;
+	else
+		qm->qp_idx++;
+
+	qp = &qm->qp_array[qm->qp_idx];
+	memzero_explicit(qp->cqe, sizeof(struct qm_cqe) * HISI_QM_Q_DEPTH);
+	qp->sq_type = sq_type;
+	qp->sq_tail = 0;
+	qp->cq_head = 0;
+	qp->cqc_phase = true;
+
+	if (qm_sqc_cfg(qp)) {
+		EMSG("Fail to set qp[%"PRIu32"] sqc", qp->qp_id);
+		goto err_proc;
```

Comment:
nitpicking: clear `qp->cqc_phase` and `qp->sq_type` upon failure?

Reply:
Not a nitpicking!
You suggestion is always helpful, We're happy to accept.